display listening ports for each service when uses curveadm status

ArberSephirotheca opened the pull request:
OSPP Project Number: 222990287

Problem Summary:
Display port number that each service is listening to by default when user uses `curveadm status` command.

### What is changed and how it works?
**internal/tui/service: status.go**
**internal/task/task/common: get_status.go**

#### get_status.go
- added additional step in `NewGetServiceStatusTask` function. New step executes pipeline command inside container to extract port numbers the service is listening to.
- If pipleline command returns any message other than port number(e.g. error message), empty string will be returned as result.

#### status.go
- added new merge function for ports: Duplicated port number will be merged, each port number is splitted by comma.
-  empty string will be return if there is no listening port.

Wine93 commented:
@ArberSephirotheca can you provide a screenshot for the `curveadm status` :)

aspirer commented:
It's better to merge your five commits to one by using `git rebase -i`.

aspirer commented:
Hi @ArberSephirotheca , thanks for your contribution.
I merged this pr and build curveadm, then run `curveadm status` command, output is like:
![image](https://user-images.githubusercontent.com/3424999/181142640-37d78711-aa24-45e6-9ecb-36d1963e7b78.png)

so we'd better to fix these problems:
1. we'd better to print service ports by using `curveadm status -v`, if verbose flag isn't set, do not print the ports
2. the ports is splited to two lines, not one, we should print them in one line.

ArberSephirotheca commented:
> 1. we'd better to print service ports by using `curveadm status -v`, if verbose flag isn't set, do not print the ports
Now the listening port will be displayed only if user enables verbose mode.
> 2. the ports is splited to two lines, not one, we should print them in one line.
Fixed.
> 3. services ports like metaserver/chunkserver/snapshotclone are not printed.
I changed the command to be executed in container when ues `curveadm status`. It now should print listening ports for all types of services.
> 4. how to deal with chunkserver replicas ports, we often have more than one chunkserver on one host, how to print them?
If user does not enable expand mode, replicas with same port number will be merged together by `port` merge function, see:
https://github.com/opencurve/curveadm/pull/123/files#diff-43002aebd9f6d4f2b2084c01747a8b1821a931f16531e9b75afed88c9c6932b2R123-R139
If expand mode is enabled, the port number for each replica is printed respectively.

aspirer commented:
@ArberSephirotheca  goodjob, but I find a little problem, the elapsed time of `status` command increased 2s (without verbose flag), we'd better fix it.
this may because we always run `lsof` commands in all service containers, even if we do not add verbose flag.
```
[curve@localhost curveadm]$ time ./sbin/curveadm status
Get Service Status: [OK]

cluster name    : mybs
cluster kind    : curvebs
cluster mds addr: 10.0.2.15:60666(leader),10.0.2.15:61666,10.0.2.15:62666

Id            Role           Host       Replica  Container Id  Status
--            ----           ----       -------  ------------  ------
3620695a3ccd  etcd           10.0.2.15  1/1      22fd65adcd06  Up 2 days
e36d51398eaf  etcd           10.0.2.15  1/1      a2b09821c5b8  Up 2 days
cb6d4957cb62  etcd           10.0.2.15  1/1      d12e3bdff012  Up 2 days
3786cb933a4d  mds            10.0.2.15  1/1      6c429be70e66  Up 15 hours
a12021be4ab1  mds            10.0.2.15  1/1      30ae577d8500  Up 3 hours
a00436fa7ff7  mds            10.0.2.15  1/1      674440f98541  Up 8 hours
aa9f2935af17  chunkserver    10.0.2.15  1/1      00bfa45fb0af  Exited (134) 36 hours ago
5b594423361d  chunkserver    10.0.2.15  1/1      1fefea2e14dc  Up 2 days
57811d7e2374  chunkserver    10.0.2.15  1/1      dd62cc23f4a8  Up 2 days
6b4804d037fa  snapshotclone  10.0.2.15  1/1      000d3a971a35  Created
9ab6612ab51e  snapshotclone  10.0.2.15  1/1      ec4a48497931  Created
5901c11a3d24  snapshotclone  10.0.2.15  1/1      96603bc6bbae  Created

real    0m3.180s
user    0m0.143s
sys     0m0.081s
[curve@localhost curveadm]$ time curveadm status
Get Service Status: [OK]

cluster name    : mybs
cluster kind    : curvebs
cluster mds addr: 10.0.2.15:60666(leader),10.0.2.15:61666,10.0.2.15:62666

Id            Role           Host       Replica  Container Id  Status
--            ----           ----       -------  ------------  ------
3620695a3ccd  etcd           10.0.2.15  1/1      22fd65adcd06  Up 2 days
e36d51398eaf  etcd           10.0.2.15  1/1      a2b09821c5b8  Up 2 days
cb6d4957cb62  etcd           10.0.2.15  1/1      d12e3bdff012  Up 2 days
3786cb933a4d  mds            10.0.2.15  1/1      6c429be70e66  Up 15 hours
a12021be4ab1  mds            10.0.2.15  1/1      30ae577d8500  Up 3 hours
a00436fa7ff7  mds            10.0.2.15  1/1      674440f98541  Up 8 hours
aa9f2935af17  chunkserver    10.0.2.15  1/1      00bfa45fb0af  Exited (134) 36 hours ago
5b594423361d  chunkserver    10.0.2.15  1/1      1fefea2e14dc  Up 2 days
57811d7e2374  chunkserver    10.0.2.15  1/1      dd62cc23f4a8  Up 2 days
6b4804d037fa  snapshotclone  10.0.2.15  1/1      000d3a971a35  Created
9ab6612ab51e  snapshotclone  10.0.2.15  1/1      ec4a48497931  Created
5901c11a3d24  snapshotclone  10.0.2.15  1/1      96603bc6bbae  Created

real    0m1.295s
user    0m0.085s
sys     0m0.056s
```

aspirer commented on internal/task/task/common/get_status.go:
```diff
@@ -89,6 +105,14 @@ func (s *step2FormatStatus) Execute(ctx *context.Context) error {
+	default:
+		exectuble = dc.GetKind() + "-" + dc.GetRole()
+	}
+	cmd := "lsof -i -P -n -sTCP:LISTEN " + "/" + dc.GetKind() + "/" + dc.GetRole() + "/sbin/" + exectuble + " | awk '{print$9}' | awk -F ':' '{print$2}'"
```
we don't need to run this command when the verbose flag isn't set, otherwise the command will become slow. this means we only get ports when verbose flag is set(curveadm status -v).

ArberSephirotheca replied:
I added an additional parameter for `NewGetServiceStatusTask` function to detect the value of verbose flag, I also change the docker exec command to `ss`. The performance has improved.

**When verbose mode is enable**:
Before(use `lsof` command):
```
curve@czy-Inspiron-5680:~/curveadm$ time ~/curveadm/sbin/curveadm status -v 
Get Service Status: [OK]  

cluster name    : czy
cluster kind    : curvefs
cluster mds addr: 127.0.0.1:6700(leader),127.0.0.1:6701,127.0.0.1:6702

Id            Role        Host       Replica  Container Id  Status      Listen Port  Log Dir                               Data Dir                            
--            ----        ----       -------  ------------  ------      -----------  -------                               --------                            
9cd171cfd2c1  etcd        127.0.0.1  1/1      41395b7765b6  Up 8 hours  23790,23800  /home/curve/curvefs/logs/etcd0        /home/curve/curvefs/data/etcd0      
6235981e6bed  etcd        127.0.0.1  1/1      e4fe54564136  Up 8 hours  23791,23801  /home/curve/curvefs/logs/etcd1        /home/curve/curvefs/data/etcd1      
b826c5034f47  etcd        127.0.0.1  1/1      37ce2674c1fb  Up 8 hours  23792,23802  /home/curve/curvefs/logs/etcd2        /home/curve/curvefs/data/etcd2      
58ca3c69f1cf  mds         127.0.0.1  1/1      1ac65ce349c0  Up 8 hours  6700,7700    /home/curve/curvefs/logs/mds0         /home/curve/curvefs/data/mds0       
c9be59c37f64  mds         127.0.0.1  1/1      1b99ab279ab4  Up 8 hours  7701         /home/curve/curvefs/logs/mds1         /home/curve/curvefs/data/mds1       
c2b6ab07776a  mds         127.0.0.1  1/1      051a0e51ce70  Up 8 hours  7702         /home/curve/curvefs/logs/mds2         /home/curve/curvefs/data/mds2       
b0b65cfea3e4  metaserver  127.0.0.1  1/1      9a85e96a0647  Up 4 hours  6800,7800    /home/curve/curvefs/logs/metaserver0  /home/curve/curvefs/data/metaserver0
eb9b3c91b4a6  metaserver  127.0.0.1  1/1      9232afa493c4  Up 4 hours  6801,7801    /home/curve/curvefs/logs/metaserver1  /home/curve/curvefs/data/metaserver1
58fd6aa8b68d  metaserver  127.0.0.1  1/1      d2cc8e04ad88  Up 4 hours  6802,7802    /home/curve/curvefs/logs/metaserver2  /home/curve/curvefs/data/metaserver2

real	0m0.854s
user	0m0.036s
sys	0m0.037s
```

After(use `ss` command):
```
curve@czy-Inspiron-5680:~/curveadm$ time ~/curveadm/sbin/curveadm status -v 
Get Service Status: [OK]  

cluster name    : czy
cluster kind    : curvefs
cluster mds addr: 127.0.0.1:6700(leader),127.0.0.1:6701,127.0.0.1:6702

Id            Role        Host       Replica  Container Id  Status      Listen Port  Log Dir                               Data Dir                            
--            ----        ----       -------  ------------  ------      -----------  -------                               --------                            
9cd171cfd2c1  etcd        127.0.0.1  1/1      41395b7765b6  Up 8 hours  23790,23800  /home/curve/curvefs/logs/etcd0        /home/curve/curvefs/data/etcd0      
6235981e6bed  etcd        127.0.0.1  1/1      e4fe54564136  Up 8 hours  23791,23801  /home/curve/curvefs/logs/etcd1        /home/curve/curvefs/data/etcd1      
b826c5034f47  etcd        127.0.0.1  1/1      37ce2674c1fb  Up 8 hours  23792,23802  /home/curve/curvefs/logs/etcd2        /home/curve/curvefs/data/etcd2      
58ca3c69f1cf  mds         127.0.0.1  1/1      1ac65ce349c0  Up 8 hours  6700,7700    /home/curve/curvefs/logs/mds0         /home/curve/curvefs/data/mds0       
c9be59c37f64  mds         127.0.0.1  1/1      1b99ab279ab4  Up 8 hours  7701         /home/curve/curvefs/logs/mds1         /home/curve/curvefs/data/mds1       
c2b6ab07776a  mds         127.0.0.1  1/1      051a0e51ce70  Up 8 hours  7702         /home/curve/curvefs/logs/mds2         /home/curve/curvefs/data/mds2       
b0b65cfea3e4  metaserver  127.0.0.1  1/1      9a85e96a0647  Up 4 hours  6800,7800    /home/curve/curvefs/logs/metaserver0  /home/curve/curvefs/data/metaserver0
eb9b3c91b4a6  metaserver  127.0.0.1  1/1      9232afa493c4  Up 4 hours  6801,7801    /home/curve/curvefs/logs/metaserver1  /home/curve/curvefs/data/metaserver1
58fd6aa8b68d  metaserver  127.0.0.1  1/1      d2cc8e04ad88  Up 4 hours  6802,7802    /home/curve/curvefs/logs/metaserver2  /home/curve/curvefs/data/metaserver2

real	0m0.565s
user	0m0.030s
sys	0m0.049s
```
**When verbose mode is disable**:
my version:
```
curve@czy-Inspiron-5680:~/curveadm$ time ~/curveadm/sbin/curveadm status 
Get Service Status: [OK]  

cluster name    : czy
cluster kind    : curvefs
cluster mds addr: 127.0.0.1:6700(leader),127.0.0.1:6701,127.0.0.1:6702

Id            Role        Host       Replica  Container Id  Status    
--            ----        ----       -------  ------------  ------    
9cd171cfd2c1  etcd        127.0.0.1  1/1      41395b7765b6  Up 8 hours
6235981e6bed  etcd        127.0.0.1  1/1      e4fe54564136  Up 8 hours
b826c5034f47  etcd        127.0.0.1  1/1      37ce2674c1fb  Up 8 hours
58ca3c69f1cf  mds         127.0.0.1  1/1      1ac65ce349c0  Up 8 hours
c9be59c37f64  mds         127.0.0.1  1/1      1b99ab279ab4  Up 8 hours
c2b6ab07776a  mds         127.0.0.1  1/1      051a0e51ce70  Up 8 hours
b0b65cfea3e4  metaserver  127.0.0.1  1/1      9a85e96a0647  Up 4 hours
eb9b3c91b4a6  metaserver  127.0.0.1  1/1      9232afa493c4  Up 4 hours
58fd6aa8b68d  metaserver  127.0.0.1  1/1      d2cc8e04ad88  Up 4 hours

real	0m0.398s
user	0m0.044s
sys	0m0.022s
```
official version:
```
curve@czy-Inspiron-5680:~/curveadm$ time curveadm status 
Get Service Status: [OK]  

cluster name    : czy
cluster kind    : curvefs
cluster mds addr: 127.0.0.1:6700(leader),127.0.0.1:6701,127.0.0.1:6702

Id            Role        Host       Replica  Container Id  Status    
--            ----        ----       -------  ------------  ------    
9cd171cfd2c1  etcd        127.0.0.1  1/1      41395b7765b6  Up 8 hours
6235981e6bed  etcd        127.0.0.1  1/1      e4fe54564136  Up 8 hours
b826c5034f47  etcd        127.0.0.1  1/1      37ce2674c1fb  Up 8 hours
58ca3c69f1cf  mds         127.0.0.1  1/1      1ac65ce349c0  Up 8 hours
c9be59c37f64  mds         127.0.0.1  1/1      1b99ab279ab4  Up 8 hours
c2b6ab07776a  mds         127.0.0.1  1/1      051a0e51ce70  Up 8 hours
b0b65cfea3e4  metaserver  127.0.0.1  1/1      9a85e96a0647  Up 4 hours
eb9b3c91b4a6  metaserver  127.0.0.1  1/1      9232afa493c4  Up 4 hours
58fd6aa8b68d  metaserver  127.0.0.1  1/1      d2cc8e04ad88  Up 4 hours

real	0m0.385s
user	0m0.033s
sys	0m0.042s
```

However, the change breaks the consistency for `NewxxxTask` function(see https://github.com/opencurve/curveadm/pull/123/commits/9f986e54b07dbd0a1cbbd8606e37e1092ef5e861#diff-c3c15742ff739f5546ec2ee2ec864e960753ea1410fe80e07a894a2c4a6eb9edL225-R233)

Please let me know if this works and if you have any suggestions or comments.

aspirer replied:
@Wine93 please check the new change about 'verbose'.

Wine93 replied:
> @Wine93 please check the new change about 'verbose'.

I will pick this feature to curveadm `v1.0`.

Wine93 commented:
LGTM!
